extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -88,24 +89,40 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
         storageDir
     );
 
-    Path outFile = new Path(String.format("%s/%s/index.zip", config.getStorageDirectory(), storageDir));
+    Path outFile = new Path(String.format(
+        "%s/%s/index.zip",
+        config.getStorageDirectory(),
+        UUIDUtils.generateUuid()
+    ));
     FileSystem fs = outFile.getFileSystem(hadoopConfig);
 
     fs.mkdirs(outFile.getParent());
     log.info("Compressing files from[%s] to [%s]", inDir, outFile);
 
     final long size;
+    final DataSegment dataSegment;
     try (FSDataOutputStream out = fs.create(outFile)) {
       size = CompressionUtils.zip(inDir, out);
+      dataSegment = createDescriptorFile(
+          segment.withLoadSpec(makeLoadSpec(outFile))
+                 .withSize(size)
+                 .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir)),
+          outFile.getParent(),
+          fs
+      );
+      Path outDir = new Path(String.format("%s/%s", config.getStorageDirectory(), storageDir));
+      if (!fs.rename(outFile.getParent(), outDir)) {
+        throw new IOException(String.format(
+            "Writing conflict detected for [%s], is there some other job running?",
+            segment.getIdentifier()
+        ));
+      }
+    }
+    finally {
+      fs.delete(outFile.getParent(), true);
```

Comment:
if it returns false then that should fail the task with the error that tempo directory couldn't be cleaned up.